crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1038,28 +1042,74 @@ impl<'db> TypeInferenceBuilder<'db> {
         definition: Definition<'db>,
     ) {
-        // TODO(dhruvmanila): Infer types from annotation or default expression
-        // TODO check that default is assignable to parameter type
-        self.infer_parameter_definition(&parameter_with_default.parameter, definition);
+        let ast::ParameterWithDefault {
+            parameter,
+            default,
+            range: _,
+        } = parameter_with_default;
+        let opt_default_ty = default
+            .as_ref()
+            .map(|default| definition_expression_ty(self.db, definition, default));
+        if let Some(annotation) = parameter.annotation.as_ref() {
+            let declared_ty = definition_expression_ty(self.db, definition, annotation);
+            let inferred_ty = if let Some(default_ty) = opt_default_ty {
+                UnionType::from_elements(self.db, [declared_ty, default_ty])
+            } else {
+                declared_ty
+            };
+            self.add_declaration_with_binding(
+                parameter.into(),
+                definition,
+                declared_ty,
+                inferred_ty,
+            );
+        } else {
+            let ty = if let Some(default_ty) = opt_default_ty {
+                UnionType::from_elements(self.db, [Type::Unknown, default_ty])
+            } else {
+                Type::Unknown
+            };
+            self.add_binding(parameter.into(), definition, ty);
+        }
     }
 
-    fn infer_parameter_definition(
+    fn infer_variadic_parameter_definition(
         &mut self,
         parameter: &ast::Parameter,
         definition: Definition<'db>,
     ) {
-        // TODO(dhruvmanila): Annotation expression is resolved at the enclosing scope, infer the
-        // parameter type from there
-        let annotated_ty = todo_type!("function parameter type");
-        if parameter.annotation.is_some() {
-            self.add_declaration_with_binding(
+        if let Some(annotation) = parameter.annotation.as_ref() {
+            let _annotated_ty = definition_expression_ty(self.db, definition, annotation);
+            // TODO tuple[annotated_ty, ...]
+            let ty = KnownClass::Tuple.to_instance(self.db);
+            self.add_declaration_with_binding(parameter.into(), definition, ty, ty);
+        } else {
+            self.add_binding(
                 parameter.into(),
                 definition,
-                annotated_ty,
-                annotated_ty,
+                // TODO tuple of Unknown
+                KnownClass::Tuple.to_instance(self.db),
             );
+        }
+    }
+
+    fn infer_variadic_keyword_parameter_definition(
+        &mut self,
+        parameter: &ast::Parameter,
+        definition: Definition<'db>,
+    ) {
+        if let Some(annotation) = parameter.annotation.as_ref() {
+            let _annotated_ty = definition_expression_ty(self.db, definition, annotation);
+            // TODO dict[str, annotated_ty]
```

Comment:
I guess I'm feeling picky today :P

```suggestion
            // TODO `dict[str, annotated_ty]`
```